drivers/serial/uart_nrfx_uart.c

```diff
@@ -291,6 +316,335 @@ static int uart_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+
+	struct uart_event evt = {
+		.type = UART_TX_ABORTED,
+		.data.tx.buf = uart0_cb.tx_buffer,
+		.data.tx.len = uart0_cb.tx_counter
+	};
+
+	user_callback(&evt);
+
+	uart0_cb.tx_buffer_length = 0;
+	uart0_cb.tx_counter = 0;
+
+	return 0;
+}
+
+static int uart_nrfx_rx_enable(struct device *dev, u8_t *buf, size_t len,
+			       u32_t timeout)
+{
+	if (uart0_cb.rx_buffer_length != 0) {
+		return -EBUSY;
+	}
+	uart0_cb.rx_enabled = 1;
+	uart0_cb.rx_buffer = buf;
+	uart0_cb.rx_buffer_length = len;
+	uart0_cb.rx_counter = 0;
+	uart0_cb.rx_secondary_buffer_length = 0;
+	uart0_cb.rx_timeout = timeout;
+
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_ERROR);
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_RXDRDY);
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_RXTO);
+	nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STARTRX);
+	nrf_uart_int_enable(uart0_addr, NRF_UART_INT_MASK_RXDRDY |
+					NRF_UART_INT_MASK_ERROR |
+					NRF_UART_INT_MASK_RXTO);
+
+	return 0;
+}
+
+static int uart_nrfx_rx_buf_rsp(struct device *dev, u8_t *buf, size_t len)
+{
+	if (!uart0_cb.rx_enabled) {
+		return -EACCES;
+	}
+	if (uart0_cb.rx_secondary_buffer_length != 0) {
+		return -EBUSY;
+	}
+	uart0_cb.rx_secondary_buffer = buf;
+	uart0_cb.rx_secondary_buffer_length = len;
+
+	return 0;
+}
+
+static int uart_nrfx_rx_disable(struct device *dev)
+{
+	if (uart0_cb.rx_buffer_length == 0) {
+		return -EFAULT;
+	}
+
+	uart0_cb.rx_enabled = 0;
+	k_delayed_work_cancel(&uart0_cb.rx_timeout_work);
+	nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STOPRX);
+
+	return 0;
+}
+
+#define UART_ERROR_FROM_MASK(mask) \
+	(mask & NRF_UART_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : mask & NRF_UART_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : mask & NRF_UART_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : mask & NRF_UART_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+
+static void rx_rdy_evt(void)
+{
+	struct uart_event event;
+	size_t rx_cnt = uart0_cb.rx_counter;
+
+	event.type = UART_RX_RDY;
+	event.data.rx.buf = uart0_cb.rx_buffer;
+	event.data.rx.len = rx_cnt - uart0_cb.rx_offset;
+	event.data.rx.offset = uart0_cb.rx_offset;
+
+	uart0_cb.rx_offset = rx_cnt;
+
+	user_callback(&event);
+}
+
+static void rx_isr(void)
+{
+	struct uart_event event;
+
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_RXDRDY);
+
+	if (!uart0_cb.rx_buffer_length || !uart0_cb.rx_enabled) {
+		/* Byte received when receiving is disabled - data lost. */
+		(void) nrf_uart_rxd_get(uart0_addr);
+	} else {
+		if (uart0_cb.rx_counter == 0) {
+			event.type = UART_RX_BUF_REQUEST;
+			user_callback(&event);
+		}
+		uart0_cb.rx_buffer[uart0_cb.rx_counter] =
+			nrf_uart_rxd_get(uart0_addr);
+		uart0_cb.rx_counter++;
+		if (uart0_cb.rx_timeout == 0) {
+			rx_rdy_evt();
+		} else {
+			k_delayed_work_submit(&uart0_cb.rx_timeout_work,
+					      uart0_cb.rx_timeout);
+		}
+	}
+
+	if (uart0_cb.rx_buffer_length == uart0_cb.rx_counter) {
+		k_delayed_work_cancel(&uart0_cb.rx_timeout_work);
```

Comment:
no, we cancel work as it may timeout even if second buffer receives no data, so work is cancelled here and submitted when we receive first byte to new buffer